lib/src/local_working_copy.rs

```diff
@@ -1162,6 +1156,7 @@ impl FileSnapshotter<'_> {
                 }
             }
         }
+        self.emit_deleted_files(&dir, file_states, &present_entries);
```

Comment:
comment (non-blocking): It's a little unfortunate how this is all structured:

- It feels like you have to very carefully sprinkle additions to `present_entries` around the functions here and think very hard about which code paths need it.
- You have to worry about the correctness of early exiting from this function without calling `emit_deleted_files`. I think it's correct for now because we only exit in the case of error, which should terminate the snapshotting altogether?
  - I guess it would be ideal if this function returned `PresentDirEntries` (and marked it as `#[must_use]`) so that it could be enforced that every code path either returned with a value that must be consumed or an error, but maybe that's a little too much ceremony in the name of type-safety.
- For example: If we start emitting warnings for too-large files rather than returning fatal errors, then we might need to accumulate the warnings but also be careful to mark them as deleted for the snapshot.

I don't have concrete ideas on how to improve it, though.

Reply:
Yeah, that's the main reason I tried to make `visit_directory()` smaller. `present_files` could be returned to caller (with `#[must_use]`), but I don't think it would make huge difference.